File: sql/util/breaches.js
// breachs.js - Paul Calvano
// Download the breaches data file from https://haveibeenpwned.com/api/v2/breaches
// 
// Usage:  wget https://haveibeenpwned.com/api/v2/breaches; node breaches.js  > breaches.tsv

const fs = require("fs");

var almanacDate = '2021-07-01';

fs.readFile("breaches", function(err, data) {
    if (err) throw err;
    const breaches = JSON.parse(data);

    console.log('date\tName\tTitle\tDomain\tBreachDate\tAddedDate\tModifiedDate\tPwnCount\tDescription\t'
    						+ 'LogoPath\tIsVerified\tIsFabricated\tIsSensitive\tIsRetired\tIsSpamList\tDataClasses');

    for (var i in breaches) {
    	console.log(
    		almanacDate
    		+ '\t' + breaches[i].Name
    		+ '\t' + breaches[i].Title
    		+ '\t' + breaches[i].Domain
    		+ '\t' + breaches[i].BreachDate
    		+ '\t' + breaches[i].AddedDate
    		+ '\t' + breaches[i].ModifiedDate
    		+ '\t' + breaches[i].PwnCount
    		+ '\t' + breaches[i].Description
    		+ '\t' + breaches[i].LogoPath
    		+ '\t' + breaches[i].IsVerified
    		+ '\t' + breaches[i].IsFabricated
    		+ '\t' + breaches[i].IsSensitive
    		+ '\t' + breaches[i].IsRetired
    		+ '\t' + breaches[i].IsSpamList
    		+ '\t' + JSON.stringify(breaches[i].DataClasses)
    	);
    }
});

File: sql/util/breaches.py
"""
1. Download breaches.json
2. Create a new table almanac.breaches_2022 via upload, with autodetected schema
3. Append the output of this query to almanac.breaches:
SELECT
    DATE('2022-06-01') AS date,
    Name,
    Title,
    Domain,
    BreachDate,
    AddedDate,
    ModifiedDate,
    PwnCount,
    Description,
    LogoPath,
    IsVerified,
    IsFabricated,
    IsSensitive,
    IsRetired,
    IsSpamList,
    TO_JSON_STRING(DataClasses) AS DataClasses
FROM
    `httparchive.almanac.breaches_2022`
"""

import json
import requests  # pylint: disable=import-error

year = 2022

with open(f"../{year}/privacy/breaches.jsonl", "w") as outfile:
    breaches = json.loads(
        requests.get("https://haveibeenpwned.com/api/v2/breaches").content
    )
    for entry in breaches:
        outfile.write(json.dumps(entry) + "\n")
